core/commands/SubprojectAddCommand.php

```diff
@@ -0,0 +1,139 @@
+<?php
+# MantisBT - A PHP based bugtracking system
+
+# MantisBT is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# MantisBT is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with MantisBT.  If not, see <http://www.gnu.org/licenses/>.
+
+require_api( 'constant_inc.php' );
+require_api( 'config_api.php' );
+require_api( 'helper_api.php' );
+require_api( 'project_api.php' );
+require_api( 'project_hierarchy_api.php' );
+
+use Mantis\Exceptions\ClientException;
+
+/**
+ * A command that adds a project subproject.
+ */
+class SubprojectAddCommand extends Command {
+	/**
+	 * $p_data['query'] is expected to contain:
+	 * - project_id (integer)
+	 *
+	 * $p_data['payload'] is expected to contain:
+	 * - subproject_id (integer)
+	 * - inherit_parent (bool)
+	 *
+	 * @param array $p_data The command data.
+	 */
+
+	/**
+	 * @var integer
+	 */
+	private $project_id;
+
+	/**
+	 * @var integer
+	 */
+	private $subproject_id;
+
+	/**
+	 * @var boolean
+	 */
+	private $inherit_parent;
+
+	/**
+	 * Constructor
+	 *
+	 * @param array $p_data The command data.
+	 */
+	function __construct( array $p_data ) {
+		parent::__construct( $p_data );
+	}
+
+	/**
+	 * Validate the data.
+	 */
+	function validate() {		
+		$this->project_id = helper_parse_id( $this->query( 'project_id' ), 'project_id' );
+
+		if( !access_has_project_level( config_get( 'manage_project_threshold' ), $this->project_id ) ) {
+			throw new ClientException(
+				'Access denied to add subprojects',
+				ERROR_ACCESS_DENIED );
+		}
+
+		if ( config_get( 'subprojects_enabled' ) == OFF ) {
+			throw new ClientException(
+				'Access denied to add subprojects',
+				ERROR_ACCESS_DENIED );
+		}
+
+		if( !project_exists( $this->project_id )) {
+			throw new ClientException(
+				"Project \"$this->project_id\"not found",
+				ERROR_PROJECT_NOT_FOUND,
+				array( $this->project_id ) );
+		}
+
+		$this->subproject_id = helper_parse_id( $this->payload( 'subproject_id' ), 'subproject_id' );
+		if( !project_exists( $this->subproject_id )) {
+			throw new ClientException(
+				"Project \"$this->subproject_id\" not found",
+				ERROR_PROJECT_NOT_FOUND,
+				array( $this->subproject_id ) );
+		}
+
+		if( $this->project_id == $this->subproject_id ) {
```

Comment:
We are better off checking this after that sub-projects is enabled and before checking access levels.  If core API does such check and loop checks, then we should remove it from here and depend on the core APIs bubbling exceptions.